PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,453 @@
+        """
+        Remove the passed modules and hence all hook scripts cached for these
+        modules from this cache.
+
+        Parameters
+        ----------
+        module_names : list
+            List of all fully-qualified module names to be removed.
+        """
+
+        for module_name in set(module_names):  # Eliminate duplicate entries.
+            if module_name in self:
+                del self[module_name]
+
+
+class ModuleHook(object):
+    """
+    Cached object encapsulating a lazy loadable hook script.
+
+    This object's `code` property provides the currently loaded contents (e.g.,
+    attributes, functions) of this hook script. The first access of this
+    property silently lazily loads this script into an in-memory code object.
+
+    Parameters
+    ----------
+    module_graph : ModuleGraph
+        Current module graph.
+    module_name : str
+        Name of the module hooked by this hook script.
+    hook_filename : str
+        Absolute or relative path of this hook script.
+    hook_module_name : str
+        Name of the in-memory module of this hook script's interpreted contents.
+    _scope : module
+        In-memory module of this hook script's interpreted contents, lazily
+        loaded on the first access of the `scope` property _or_ `None` if this
+        property has yet to be accessed.
+    """
+
+    ## Private methods
+
+    def __init__(
+        self,
+        module_graph,
+        module_name,
+        hook_filename,
+        hook_module_name_prefix,
+    ):
+        """
+        Initialize this metadata.
+
+        Parameters
+        ----------
+        module_graph : ModuleGraph
+            Current module graph.
+        module_name : str
+            Name of the module hooked by this hook script.
+        hook_filename : str
+            Absolute or relative path of this hook script.
+        hook_module_name_prefix : str
+            String prefixing the name of the in-memory module for this hook
+            script's code object. To avoid namespace clashes with similar
+            modules created by other `ModuleHook` objects in other
+            `ModuleHookCache` containers, this string _must_ be unique to the
+            `ModuleHookCache` container containing this `ModuleHook` object. If
+            this string is non-unique, an existing in-memory module will be
+            erroneously reused when lazily loading this hook script, thus
+            erroneously resanitizing previously sanitized hook script attributes
+            (e.g., `datas`) with the `format_binaries_and_datas()` helper.
+        """
+
+        # Note that the passed module graph is already a weak reference,
+        # avoiding circular reference issues. See ModuleHookCache.__init__().
+        self.module_graph = module_graph
+        self.module_name = module_name
+        self.hook_filename = hook_filename
+
+        # Name of the in-memory module fabricated to refer to this hook script.
```

Comment:
Puh! What a debug orgy. This is really driving one nuts. Thanks a lot for this long explanation. 

But I see a chance of getting rid of it. You say, this is caused by data already sanitized by format_binaries_and_datas() being in the hook. Now after eliminating the `scope` namespace and implementing `__getattr__` (see [here](https://github.com/pyinstaller/pyinstaller/pull/1858#issuecomment-194051310)), the sanitized data is not pushed into the hook's namespace and thus this prblem should not occur anymore. Could be worth trying.